File: app/src/main/java/com/googlecode/tesseract/android/NativeBinding.java
package com.googlecode.tesseract.android;

import com.googlecode.leptonica.android.Pix;
import com.renard.ocr.R;

import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.util.Log;

public class NativeBinding {


    public interface ProgressCallBack {
        void onProgressImage(long nativePix);

        void onProgressText(@StringRes int message);

        void onLayoutAnalysed(long nativePixaText, long nativePixaImages);
    }

    private static final String TAG = "NativeBinding";

    static {
        System.loadLibrary("pngo");
        System.loadLibrary("lept");
        System.loadLibrary("tess");
        System.loadLibrary("image_processing_jni");
        nativeClassInit();
    }

    @Nullable
    private ProgressCallBack mCallBack;

     //Used by the native code
    @SuppressWarnings("unused")
    private long mNativeData;


    public synchronized void setProgressCallBack(@Nullable  ProgressCallBack callBack) {
        mCallBack = callBack;
    }

    public NativeBinding() {
        nativeConstruct();
    }

    public void destroy() {
        Log.d(TAG, "destroying native data");
        nativeDestruct();
    }

    public long[] combinePixa(long nativePixaText, long nativePixaImages, int[] selectedTexts, int[] selectedImages) {
        return combineSelectedPixa(nativePixaText, nativePixaImages, selectedTexts, selectedImages);
    }

    public void analyseLayout(Pix pixs) {
        nativeAnalyseLayout(pixs.getNativePix());
    }

    public long convertBookPage(Pix pixs) {
        return nativeOCRBook(pixs.getNativePix());
    }

    /**
     * called from native code
     */
    @SuppressWarnings("unused")
    private synchronized void onProgressImage(final long nativePix) {
        if (mCallBack != null) {
            mCallBack.onProgressImage(nativePix);
        }
    }

    /**
     * called from native
     * static const int MESSAGE_IMAGE_DETECTION = 0; static const int
     * MESSAGE_IMAGE_DEWARP = 1; static const int MESSAGE_OCR = 2; static const
     * int MESSAGE_ASSEMBLE_PIX = 3; static const int MESSAGE_ANALYSE_LAYOUT =
     * 4;
     */
    @SuppressWarnings("unused")
    private void onProgressText(int id) {
        int messageId = 0;
        switch (id) {
            case 0:
                messageId = R.string.progress_image_detection;
                break;
            case 1:
                messageId = R.string.progress_dewarp;
                break;
            case 2:
                messageId = R.string.progress_ocr;
                break;
            case 3:
                messageId = R.string.progress_assemble_pix;
                break;
            case 4:
                messageId = R.string.progress_analyse_layout;
                break;

        }
        if (messageId != 0 && mCallBack!=null) {
            mCallBack.onProgressText(messageId);
        }
    }

    @SuppressWarnings("unused")
    private void onLayoutElements(long nativePixaText, long nativePixaImages) {
        if(mCallBack!=null){
            mCallBack.onLayoutAnalysed(nativePixaText, nativePixaImages);
        }
    }


    private static native void nativeClassInit();

    private native void nativeConstruct();

    private native void nativeDestruct();

    private native long nativeOCRBook(long nativePix);

    private native long[] combineSelectedPixa(long nativePixaTexts, long nativePixaImages, int[] selectedTexts, int[] selectedImages);

    private native long nativeAnalyseLayout(long nativePix);

}

File: app/src/main/java/com/googlecode/tesseract/android/NativeBinding.kt
package com.googlecode.tesseract.android

import android.support.annotation.StringRes
import android.util.Log
import com.googlecode.leptonica.android.Pix
import com.renard.ocr.R

class NativeBinding {

    private var mCallBack: ProgressCallBack? = null

    //Used by the native code
    private val mNativeData: Long = 0


    interface ProgressCallBack {
        fun onProgressImage(nativePix: Long)

        fun onProgressText(@StringRes message: Int)

        fun onLayoutAnalysed(nativePixaText: Long, nativePixaImages: Long)
    }


    fun setProgressCallBack(callBack: ProgressCallBack?) {
        mCallBack = callBack
    }

    init {
        nativeConstruct()
    }

    fun destroy() {
        Log.d(TAG, "destroying native data")
        nativeDestruct()
    }

    fun combinePixa(nativePixaText: Long, nativePixaImages: Long, selectedTexts: IntArray, selectedImages: IntArray): LongArray {
        return combineSelectedPixa(nativePixaText, nativePixaImages, selectedTexts, selectedImages)
    }

    fun analyseLayout(pixs: Pix) {
        nativeAnalyseLayout(pixs.nativePix)
    }

    fun convertBookPage(pixs: Pix): Long {
        return nativeOCRBook(pixs.nativePix)
    }

    /**
     * called from native code
     */
    @Synchronized
    private fun onProgressImage(nativePix: Long) {
        if (mCallBack != null) {
            mCallBack!!.onProgressImage(nativePix)
        }
    }

    /**
     * called from native
     * static const int MESSAGE_IMAGE_DETECTION = 0; static const int
     * MESSAGE_IMAGE_DEWARP = 1; static const int MESSAGE_OCR = 2; static const
     * int MESSAGE_ASSEMBLE_PIX = 3; static const int MESSAGE_ANALYSE_LAYOUT =
     * 4;
     */
    private fun onProgressText(id: Int) {
        var messageId = 0
        when (id) {
            0 -> messageId = R.string.progress_image_detection
            1 -> messageId = R.string.progress_dewarp
            2 -> messageId = R.string.progress_ocr
            3 -> messageId = R.string.progress_assemble_pix
            4 -> messageId = R.string.progress_analyse_layout
        }
        if (messageId != 0 && mCallBack != null) {
            mCallBack!!.onProgressText(messageId)
        }
    }

    /**
     * called from native
     */
    private fun onLayoutElements(nativePixaText: Long, nativePixaImages: Long) {
        if (mCallBack != null) {
            mCallBack!!.onLayoutAnalysed(nativePixaText, nativePixaImages)
        }
    }

    private external fun nativeConstruct()

    private external fun nativeDestruct()

    private external fun nativeOCRBook(nativePix: Long): Long

    private external fun combineSelectedPixa(nativePixaTexts: Long, nativePixaImages: Long, selectedTexts: IntArray, selectedImages: IntArray): LongArray

    private external fun nativeAnalyseLayout(nativePix: Long)

    companion object {

        private val TAG = "NativeBinding"

        init {
            nativeClassInit()
        }


        @JvmStatic
        private external fun nativeClassInit()
    }

}
